### ISSUE TITLE
[$500] Report gets blank after resizing

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#31503]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.0.3
**Reproducible in staging?:** y 
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @shubham1206agra 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1700236291136659

## Action Performed:

(watch the vid before you try to reproduce, it'll be more clear)
1. Open any report in full size
2. Go to small screen view
3. Press back header
4. Go to large view (>800) again

## Expected Result:
All the messages should be loading

## Actual Result:
Report becomes blank

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

![Screenshot 2023-11-17 at 9 19 23 PM](https://github.com/Expensify/App/assets/38435837/5d57ecec-4bcc-43fe-9e20-f7f4cd73f130)

https://github.com/Expensify/App/assets/38435837/d219b2bc-242c-4580-87bc-2a025cb7bf8c

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0195fdcb43d1c84e8f</li>
        <li>Upwork Job ID: 1725605698649042944</li>
        <li>Last Price Increase: 2023-11-24</li>
<li>Automatic offers: </li>
<ul>
<li>alitoshmatov | Reviewer | 27909873</li>
<li>tienifr | Contributor | 27909877</li>
</ul></li>
    </ul>
</details>

**Comment by mallenexpensify** [comment]
Was able to reproduce after watching vid, def something we want t ofix.  I _feel_ like a version of this bug has existed before, I couldn't find any GHs for it though.

**Comment by barros001** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When on a report, and switching screen to the small view (mobile), clicking back on the report (which takes you to LHN), and then switching back to desktop view, central panel will be empty. The same happens if you load the app on mobile view (which will land you on LHN) and expanding the window to desktop size, no central panel will be loaded.

### What is the root cause of that problem?

The root problem is is the sequence in which things are rendered. The app is structured this way (very simplified)

```
<NavigatorRoot>
  <AuthScreens>
    <Navigator>...</Navigator>
  </AuthScreens>
</NavigatorRoot>
```

When you switch from mobile to desktop (and vice versa), these components would all re-render because they all relay on `isSmallScreenWidth`, which comes from `useWindowDimensions` hook. Normally things would happen in the correct sequence, but when switching views, things get out of order. Here's why:

https://github.com/Expensify/App/blob/3f7e66a8e39b8ecfe4921c74f11f168deafd5d21/src/libs/Navigation/NavigationRoot.js#L74-L80

When `isSmallScreenWidth`, this hook gets executed first (as it sits at the top of the tree), which in turn will call `navigationRef.resetRoot(navigationRef.getRootState());`. That will trigger [getRehydratedState](https://github.com/Expensify/App/blob/3f7e66a8e39b8ecfe4921c74f11f168deafd5d21/src/libs/Navigation/AppNavigator/createCustomStackNavigator/CustomRouter.js#L70-L80) which will look for `isSmallScreenWidth`. At that point in time, `Navigator` hasn't re-rendered yet and the value of `isSmallScreenWidth` is still the previous value (still true in this case). For that reason, it will not add the central panel.

### What changes do you think we should make in order to solve the problem?

We need to delay calling `navigationRef.resetRoot(navigationRef.getRootState());` until everything have been re-rendered. To accomplish that I propose the following change:

```js
    useEffect(() => {
        if (!navigationRef.isReady() || !props.authenticated) {
            return;
        }
        // We need to force state rehydration so the CustomRouter can add the CentralPaneNavigator route if necessary.
        setTimeout(() => {
            navigationRef.resetRoot(navigationRef.getRootState());
        }, 0);
    }, [isSmallScreenWidth, props.authenticated]);
```

Wrapping the call to `resetRoot` in a `setTimeout` with a 0 delay will just let javascript process everything in the queue first, which means it will let react re-render everything before going back to it. This solves the problem and the central panel is added back.


### What alternative solutions did you explore? (Optional)


**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by derekops333** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Report becomes blank when switching from desktop to mobile view, pressing back button and then going back to the desktop view

### What is the root cause of that problem?

`getRehydratedState` function inside the `CustomRouter` is not called immediately after we switch from mobile to desktop (and other way around)


### What changes do you think we should make in order to solve the problem?

In the `ResponsiveStackNavigator`, we should add simple `useEffect` that is triggered when `isSmallScreenWidth` is changed and call `navigation.reset(state);` This will manually call `getRehydratedState` every time we have a transition  desktop-mobile or mobile-desktop transition.

**Comment by ayazalavi** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Upon resize windows from large -> small -> large size in width, central pane navigator doesnt gets loaded.

### What is the root cause of that problem?
Root cause of the problem is that CustomRouter below is not getting rehydrated based on screen size changes

https://github.com/Expensify/App/blob/4df083c10f6f9c8ba0aaaabcdba86817246d6bbc/src/libs/Navigation/AppNavigator/createCustomStackNavigator/CustomRouter.js#L65

the variable getting used to monitor screen size change is react state variable 

https://github.com/Expensify/App/blob/4df083c10f6f9c8ba0aaaabcdba86817246d6bbc/src/libs/Navigation/AppNavigator/createCustomStackNavigator/index.js#L50

### What changes do you think we should make in order to solve the problem?

Instead of entirely depending upon getRehydratedState for adding central pane we can use getStateForAction which can be triggered based on certain condition e.g. screen size changed

So basically as soon as screen size changes we can call following code in 
https://github.com/Expensify/App/blob/4df083c10f6f9c8ba0aaaabcdba86817246d6bbc/src/libs/Navigation/AppNavigator/createCustomStackNavigator/index.js#L49

navigation.dispatch({
            type: 'SCREEN_CHANGED',
            // Additional payload or parameters
          });

and in CustomRouter add something like this besides getRehydratedState

https://github.com/Expensify/App/blob/3f7e66a8e39b8ecfe4921c74f11f168deafd5d21/src/libs/Navigation/AppNavigator/createCustomStackNavigator/CustomRouter.js#L65-L70

 getStateForAction(state, action, options) {
            // eslint-disable-next-line default-case
            switch (action.type) {
                case "SCREEN_CHANGED": {
                    if (!isAtLeastOneCentralPaneNavigatorInState(state)) {
                        // If we added a route we need to make sure that the state.stale is true to generate new key for this route
                        // eslint-disable-next-line no-param-reassign
                        Navigation.navigate(ROUTES.REPORT_WITH_ID.getRoute(getTopMostReportIDFromRHP(state)));
                    }
                    return state                    
                }
            }
            return stackRouter.getStateForAction(state, action, options);
          },

As per my testing this fixes the issue.
### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by alitoshmatov** [comment]
It is the same as https://github.com/Expensify/App/issues/28354. @tienifr, since you worked on previous issue and have proposal here, can explain why this issue resurfaced.

**Comment by mallenexpensify** [comment]
Thanks @tsa321 for the link. 
Also tagging @narefyev91 and @arosiclair who worked on 
- https://github.com/Expensify/App/issues/28354

**Comment by ayazalavi** [comment]
Hey @alitoshmatov, about the issue you brought up: first, we tried sorting it out with useWindowDimensions, but it still caused some hiccups. The trouble really roots in our reliance on getRehydratedState, which is an internal callback in React Navigation. This callback kicks off the addition of a central pane, but we can't directly call it ourselves. So, for any tweaks we need after navigation loads, we've gotta find a different way.

I suggested using another approach to reach that central pane when state changes. I'd love to hear your take on this. @narefyev91 and @arosiclair, your thoughts matter too. What do you all think?

**Comment by derekops333** [comment]
> Hey @alitoshmatov, about the issue you brought up: first, we tried sorting it out with useWindowDimensions, but it still caused some hiccups. The trouble really roots in our reliance on getRehydratedState, which is an internal callback in React Navigation. This callback kicks off the addition of a central pane, but we can't directly call it ourselves. So, for any tweaks we need after navigation loads, we've gotta find a different way.
> 
> I suggested using another approach to reach that central pane when state changes. I'd love to hear your take on this. @narefyev91 and @arosiclair, your thoughts matter too. What do you all think?

My solution does manually call `getRehydratedState` on screen size change (mobile-desktop-mobile). It resets the state (but actually keeping the same state parameter) which basically just re-runs `getRehydratedState`

**Comment by derekops333** [comment]
@m-natarajan is there any updates related to this issue?

**Comment by alitoshmatov** [comment]
> What do you think? @alitoshmatov

Okay, I also couldn't find why this issue occurred again. The RCA you provided is correct. 

Reviewing proposals.

**Comment by alitoshmatov** [comment]
@tienifr 's [proposal](https://github.com/Expensify/App/issues/31503#issuecomment-1817099821) looks straightforward and solve the issue.

C+ reviewed 🎀 👀 🎀

**Comment by alitoshmatov** [comment]
@barros001 Your RCA is correct. But I don't think moving useEffect into `CustomRouter` is a good idea since it is not related to CustomRouter.

**Comment by ayazalavi** [comment]
@alitoshmatov proposal you selected is only hypothetical and it will cause more complexity since it is already tried before. my solution is just 2-3 lines of code and is quite easier to understand as well..

**Comment by ayazalavi** [comment]
code will be duplicated as well for calculating screen size. this is another problem in @tienifr proposal.

**Comment by ayazalavi** [comment]
https://github.com/Expensify/App/assets/767439/6d67972b-4d96-46ae-8dad-303987504207

**Comment by Li357** [comment]
@mallenexpensify Can you reapply the engineering label to get another internal engineer here? I'll be back at school full-time EOW, all the best!

**Comment by Gonals** [comment]
@alitoshmatov. There seems to still be some discussion ongoing. Any thoughts or do you stand by the selected proposal?

**Comment by alitoshmatov** [comment]
I am reconsidering my decision. 

I do agree with @barros001 's point, his [proposal](https://github.com/Expensify/App/issues/31503#issuecomment-1817167541)  also solve the issue  

But I am preferring @tienifr  's [proposal](https://github.com/Expensify/App/issues/31503#issuecomment-1817099821) which suggests simpler approach for solving this issue

@Gonals what do you think?

**Comment by alitoshmatov** [comment]
@ayazalavi I still think your solution is over complicating things by adding additional logic

**Comment by barros001** [comment]
> I do agree with @barros001 's point, his https://github.com/Expensify/App/issues/31503#issuecomment-1817167541 also solve the issue

It's a very simple solution to the problem, without added complexity and without duplicating logic (replicating how `isSmallScreenWidth` is calculated).

**Comment by derekops333** [comment]
@alitoshmatov could you please review my proposal?

**Comment by shubham1206agra** [comment]
@alitoshmatov I think we should aim for race condition correction here. And replicating `isSmallScreenWidth` condition is dangerous as we have different implementation for web and native devices, due to iPad, and we will aim to fix this in the future.

If you still confused, we should take this conversation to slack to get more 👀

cc @Gonals

**Comment by shubham1206agra** [comment]
@tienifr Please see `useWindowDimensions` implementation

**Comment by ayazalavi** [comment]
> @ayazalavi I still think your solution is over complicating things by adding additional logic

without disturbing current logic I think my solution provides more robust solution saving us from adding additional knots in the same code.

**Comment by Gonals** [comment]
> I am reconsidering my decision.
> 
> I do agree with @barros001 's point, his [proposal](https://github.com/Expensify/App/issues/31503#issuecomment-1817167541) also solve the issue
> 
> But I am preferring @tienifr 's [proposal](https://github.com/Expensify/App/issues/31503#issuecomment-1817099821) which suggests simpler approach for solving this issue
> 
> @Gonals what do you think?

Taking a look at the proposals, I think I agree. Moving forward with @tienifr's proposal as you originally suggested 👍

**Comment by alitoshmatov** [comment]
Issue assigned to @tienifr waiting for the PR

**Comment by alitoshmatov** [comment]
BugZero Checklist: The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

> The PR that introduced the bug has been identified. Link to the PR: 

https://github.com/Expensify/App/pull/28683
> The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

https://github.com/Expensify/App/pull/28683#discussion_r1429643257

> A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: 

No need for a discussion

> Determine if we should create a regression test for this bug.

No need for a regression test

**Comment by alitoshmatov** [comment]
This one is ready for a payment

cc: @mallenexpensify

**Comment by dangrous** [comment]
Hi all! The [solution](https://github.com/Expensify/App/pull/32422) to this caused [iPad to break](https://github.com/Expensify/App/issues/34383) and so we [reverted it](https://github.com/Expensify/App/pull/34396). So - we need a new solution to this bug. Reopening for that. cc @mountiny and @AndrewGable

**Comment by mountiny** [comment]
@mallenexpensify @alitoshmatov @tienifr Could you first confirm if the original bug is present on main/ staging now? then we can export again and seek proposals

**Comment by shubham1206agra** [comment]
@mountiny Confirming the bug repro. Please make this deploy blocker.

> @alitoshmatov I think we should aim for race condition correction here. And replicating `isSmallScreenWidth` condition is dangerous as we have different implementation for web and native devices, due to iPad, and we will aim to fix this in the future.
> 
> If you still confused, we should take this conversation to slack to get more 👀
> 
> cc @Gonals

Looks like I have warned them about this in advance. But they ignored this :(.

@mountiny I can implement a quick fix for this. Or we can try to implement @barros001's [proposal](https://github.com/Expensify/App/issues/31503#issuecomment-1817167541), which will fix this issue and will not introduce special iPad logic.

**Comment by mountiny** [comment]
Making this External again, I think the issue is edge case now to be a deploy blocker

**Comment by alitoshmatov** [comment]
I am just wondering how exactly our solution crashed ipad app. I have tried to reproduce iPad issue before revert and got this error:

https://github.com/Expensify/App/assets/59907218/05fa893d-40ce-42d5-8fc2-db4130518acd

<img width="588" alt="Screenshot 2024-01-15 at 01 20 57" src="https://github.com/Expensify/App/assets/59907218/d9dccf40-d945-4e4f-9ec9-52428675caa5">


@shubham1206agra You mentioned race condition, can you elaborate on that if you think that was the root cause of iPad issue

**Comment by shubham1206agra** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.

Report gets blank after resizing

### What is the root cause of that problem?

The `getRehydratedState` [here](https://github.com/Expensify/App/blob/3f7e66a8e39b8ecfe4921c74f11f168deafd5d21/src/libs/Navigation/AppNavigator/createCustomStackNavigator/CustomRouter.js#L70) is run before the `useWindowDimensions` hook render again, leading to the `isSmallScreenWidth` is still true when `getRehydratedState` is run while it actually should be false. This causes the Central Pane not to be added when resizing the screen and the report is blank.

### What changes do you think we should make in order to solve the problem?

In [here](https://github.com/Expensify/App/blob/3f7e66a8e39b8ecfe4921c74f11f168deafd5d21/src/libs/Navigation/AppNavigator/createCustomStackNavigator/CustomRouter.js#L70), calculate the `isSmallScreenWidth` by using the `Dimensions.get('window').width` inside `getRehydratedState` to get the latest value.

```js
const windowWidth = Dimensions.get('window').width;
const isSmallScreenWidth = windowWidth <= variables.mobileResponsiveWidthBreakpoint;
```

The above code is for the web platforms. For native devices, the `isSmallScreenWidth` is always true from https://github.com/Expensify/App/pull/27004 to fix the iPad layout for the time being as we don't officially fully support iPad / tablet devices.

The logic is the same as the existing `isSmallScreenWidth` calculation [here](https://github.com/Expensify/App/blob/3f7e66a8e39b8ecfe4921c74f11f168deafd5d21/src/hooks/useWindowDimensions/index.ts#L14), just that it uses current value of the window dimension instead of using a hook, and should be used in cases like this the `isSmallScreenWidth` needs to be retrieved immediately without delay of a hook render. The logic can be extracted to a `getIsSmallScreenWidth` method for reuse.

Then we can clean up the existing `getIsSmallScreenWidth` [here](https://github.com/Expensify/App/blob/3f7e66a8e39b8ecfe4921c74f11f168deafd5d21/src/libs/Navigation/AppNavigator/createCustomStackNavigator/index.js#L61) in web since it's not used any more

### What alternative solutions did you explore? (Optional)

NA

**Comment by shubham1206agra** [comment]
> @shubham1206agra You mentioned race condition, can you elaborate on that if you think that was the root cause of iPad issue

Nope, this is just iPad behavior got out of sync with useWindowDimensions. 

I can raise the new PR with original solution with minor fixes for iPad. (This is mostly 2-3 lines addition).

**Comment by mountiny** [comment]
@alitoshmatov I also believe you might have to try on larger screen ipad

**Comment by mallenexpensify** [comment]
This can be closed... right?  
[Above I see payment has already been made ](https://github.com/Expensify/App/issues/31503#issuecomment-1861224354) and, if there was a regression (which it looks like there was) it was for an unsupported device, iPad, so I don't think we'd reduce pay for that

**Comment by shubham1206agra** [comment]
@mallenexpensify We reverted the change so its still broken. @alitoshmatov needs to review the proposal [here](https://github.com/Expensify/App/issues/31503#issuecomment-1892192186) to fix this again. So can you reopen this again?

**Comment by dangrous** [comment]
Yep, you're right, reopening as we reverted the PR that fixed it.

We should determine whether @tienifr and @alitoshmatov should work to fix (again) since they were the original assignees, or whether we want to open it up again.

Thanks!

**Comment by alitoshmatov** [comment]
I am a little bit confused. @shubham1206agra your proposal is exactly the same as @tienifr 's, Are you suggesting that reimplementing this solution is correct approach. If yes, I think I and @tienifr can reimplement this easily since it is the same old solution. I am happy to review it without any payment.

**Comment by shubham1206agra** [comment]
Reimplementing with little changes. But let's wait for Ideal Nav to stabilize to check the repro of the issue.

**Comment by alitoshmatov** [comment]
@shubham1206agra That looks like a plan, I do agree with holding this issue since this issue is an edge case

**Comment by mallenexpensify** [comment]
Posted in the Wave 8 room. to check

**Comment by mallenexpensify** [comment]
I'm going to close this. Coming from the [internal slack discussion](https://expensify.slack.com/archives/C036QM0SLJK/p1707867617551479)
> We don’t officially support re-sizing to my knowledge. We had a bunch of issues open in the repo recently and closed them. I don’t think this is a priority right now.

If/when we want to address resizing we'll be able to easily find this issue cuz `resizing` is in the title.
